doc/design/zarr-support-3.md

```diff
@@ -143,6 +143,54 @@ This bubbles up to the top of the zarr archive, where the final `.checksum` for
 
 No spaces are used in JSON encodings.
 
+## Calculating checksums
+Experience has proven that updating the .checksum file in the upload complete request takes too long.
+Therefore, updating the `.checksum` file happens in a celery task.
+The upload complete endpoint will enqueue updates to perform, which are then handled as soon as possible by an arbitrary number of celery workers.
+This allows us to scale the number of celery workers so that the checksum for a zarr archive is available almost immediately after upload is complete.
+
+### ZarrChecksumUpdate
+The `ZarrChecksumUpdate` model is used to store updates that need to be applied to a zarr archive.
+It's fields will include:
+* The `ZarrArchive` it applies to.
+* The `parent_directory` containing the `.checksum` file being modified.
+* The `name` of the file being changed in the `parent_directory`.
+* An `update_type` enum:
+  * `FILE`: adding or updating a file checksum
+  * `DIRECTORY`: adding or updating a directory checksum
+  * `REMOVAL`: removing a checksum
+* The `depth` of the update; basically the number of slashes in `parent_directory`.
+* A `status` enum:
+  * `PENDING`: not yet picked up by a worker
+  * `IN_PROGRESS`: being processed by a worker
+
+
+In the upload complete endpoint, a new `ZarrChecksumUpdate` is created with an `update_type` of `FILE` for every path.
+In the file delete endpoint, a new `ZarrChecksumUpdate` is created with an `update_type` of `REMOVAL` for every path.
+Both endpoints will then dispatch a single checksum update task.
+
+### Checksum update task
+The checksum update task takes a `zarr_id` as an argument.
+It will then run the following loop:
+* Select the `ZarrChecksumUpdate` with the greatest `depth` and status `PENDING`.
+  * If there is no such `ZarrChecksumUpdate`, terminate.
+* Select for update all `ZarrChecksumUpdate` with the same `parent_directory`.
+  These represent all the updates waiting to be applied to a `.checksum` file.
+  * If the select for update failed, another worker selected the same `parent_directory` in a race condition.
+    Catch the error and terminate silently.
+* Set the `status` of all selected `ZarrChecksumUpdates` to `IN_PROGRESS`.
+* Dispatch another checksum update task.
```

Comment:
Only dispatch on the first loop to prevent exponential growth of the task queue.